### ISSUE TITLE
Querying "enabled" from the log crate makes tracing swallow the next event 

### ISSUE THREAD

**Issue by rtimush** [tokio-rs/tracing#2519]
### Version

```
├── tracing v0.1.37
│   ├── tracing-attributes v0.1.23 (proc-macro)
│   └── tracing-core v0.1.30
└── tracing-subscriber v0.3.16
    ├── tracing v0.1.37 (*)
    ├── tracing-core v0.1.30 (*)
    └── tracing-log v0.1.3
        └── tracing-core v0.1.30 (*)
```
### Platform

```
Darwin roman.timushev-MacBookPro 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000 arm64 arm Darwin
```

### Description

Querying "enabled" from the `log` crate in some cases makes `tracing` swallow the next event.
Example code:
```rust
use tracing_subscriber::{fmt, prelude::*, EnvFilter};

fn main() {
    tracing_subscriber::registry()
        .with(fmt::layer().with_filter(EnvFilter::from_default_env()))
        .init();

    tracing::info!("this is logged");
    let _ = log::logger().enabled(&log::Metadata::builder().level(log::Level::Debug).build());
    tracing::info!("this is not logged");
    tracing::info!("this is logged again");
}
```

If executed with `RUST_LOG=<crate_name>=debug,info`, the second `info!` is not logged.

This doesn't happen if:
- the filter is applied not to the logging layer but to the whole subscriber
- RUST_LOG is set to `debug` or `<crate_name>=info,info` 

The Cargo.toml dependencies section:
```TOML
[dependencies]
log = "0.4.17"
tracing = "0.1.37"
tracing-subscriber = { version = "0.3.16", features = ["env-filter"] }
```

**Comment by scottlamb** [comment]
Thanks for filing this! I was on the trail of why one of my spans getting swallowed, realized commenting out `tracing_log::LogTracer::init()` "fixed" it, searched, and found this timely bug report with a nice minimal repro.

**Comment by scottlamb** [comment]
I'd love to get this fixed, and have been reading the code a bit, but I'm really confused about how per-layer filtering is supposed to work. E.g., starting from the basics:

https://github.com/tokio-rs/tracing/blob/a0126b2e2d465e8e6d514acdf128fcef5b863d27/tracing-subscriber/src/registry/sharded.rs#L229-L232

`<Registry as Subscriber>::enabled` returns some pre-set state from a thread local. Is there some contract that something else should have been called first? From reading [tracing::Subscriber docs](https://docs.rs/tracing/0.1.37/tracing/subscriber/trait.Subscriber.html#tymethod.enabled), it sounds like it's more common to go through `register_callsite`, but this isn't actually required? So I don't understand how this `enabled` implementation can work? am I missing something?

**Comment by hds** [comment]
I've reproduced this without log. It seems that `Dispatch::enabled()` affects the next call-site to be registered.

Cargo.toml:

```toml
[package]
name = "tracing-lost-callsite"
version = "0.1.0"
edition = "2021"

[dependencies]
tracing = "0.1.37"
tracing-core = "0.1.31"
tracing-subscriber = "0.3.17"
```

src/main.rs:

```rust
fn main() {
    use tracing_subscriber::prelude::*;

    // Change this to DEBUG and the supressed INFO event will be recorded.
    let filter = tracing_core::LevelFilter::INFO;
    tracing_subscriber::registry()
        .with(tracing_subscriber::fmt::layer().with_filter(filter))
        .init();

    tracing::info!("this event is unaffected");

    let metadata = make_metadata_with_level(tracing_core::Level::DEBUG);
    let enabled = tracing::dispatcher::get_default(|dispatch| dispatch.enabled(&metadata));

    tracing::info!("this event gets suppressed");
    tracing::info!("The callsite was enabled? {}", enabled);
}

fn make_metadata_with_level(level: tracing_core::Level) -> tracing::Metadata<'static> {
    struct Cs;
    impl tracing_core::Callsite for Cs {
        fn set_interest(&self, _interest: tracing_core::Interest) {}
        fn metadata(&self) -> &tracing_core::Metadata<'_> {
            unimplemented!()
        }
    }

    tracing::Metadata::new(
        "",
        "app",
        level,
        None,
        None,
        None,
        tracing::field::FieldSet::new(&[], tracing_core::identify_callsite!(&Cs)),
        tracing::metadata::Kind::EVENT,
    )
}
```

**Comment by hawkw** [comment]
> I've reproduced this without log. It seems that `Dispatch::enabled()` affects the next call-site to be registered.
> 
> Cargo.toml:
> 
> ```toml
> [package]
> name = "tracing-lost-callsite"
> version = "0.1.0"
> edition = "2021"
> 
> [dependencies]
> tracing = "0.1.37"
> tracing-core = "0.1.31"
> tracing-subscriber = "0.3.17"
> ```
> 
> src/main.rs:
> 
> ```rust
> fn main() {
>     use tracing_subscriber::prelude::*;
> 
>     // Change this to DEBUG and the supressed INFO event will be recorded.
>     let filter = tracing_core::LevelFilter::INFO;
>     tracing_subscriber::registry()
>         .with(tracing_subscriber::fmt::layer().with_filter(filter))
>         .init();
> 
>     tracing::info!("this event is unaffected");
> 
>     let metadata = make_metadata_with_level(tracing_core::Level::DEBUG);
>     let enabled = tracing::dispatcher::get_default(|dispatch| dispatch.enabled(&metadata));
> 
>     tracing::info!("this event gets suppressed");
>     tracing::info!("The callsite was enabled? {}", enabled);
> }
> 
> fn make_metadata_with_level(level: tracing_core::Level) -> tracing::Metadata<'static> {
>     struct Cs;
>     impl tracing_core::Callsite for Cs {
>         fn set_interest(&self, _interest: tracing_core::Interest) {}
>         fn metadata(&self) -> &tracing_core::Metadata<'_> {
>             unimplemented!()
>         }
>     }
> 
>     tracing::Metadata::new(
>         "",
>         "app",
>         level,
>         None,
>         None,
>         None,
>         tracing::field::FieldSet::new(&[], tracing_core::identify_callsite!(&Cs)),
>         tracing::metadata::Kind::EVENT,
>     )
> }
> ```

Thanks for the repro, this is great. We could add this as a regression test for a fix.

Is constructing a `Metadata` with `None` location, filename, and module path necessary to trigger this issue?

**Comment by hds** [comment]
@hawkw I don't believe so, this was just supposed to be minimal. Another repro is via `tracing-log`. This is the one described in this issue, but I don't know whether the `Metadata` created from `tracing-log` has no location, filename, etc.

```rust
use tracing_subscriber::prelude::*;

fn main() {
    let env_filter = tracing_subscriber::filter::EnvFilter::new("other_mod=debug,debug");
    let fmt_layer = tracing_subscriber::fmt::layer().with_filter(env_filter);
    tracing_subscriber::registry().with(fmt_layer).init();

    tracing::error!("this error is unaffected");
    _ = log::log_enabled!(target: "other_mod", log::Level::Debug);
    tracing::error!("this error gets suppressed");
}
```

**Comment by smalis-msft** [comment]
I believe I'm hitting this issue without log as I have code that is manually calling enabled!, not logging an event, has a Filtered in the stack, and am seeing disappearing events. Could the title of the issue be changed so that the mention of the log crate isn't included, since it doesn't seem to be needed for this to repro?

**Comment by lennartkloock** [comment]
Is there an ETA for this? Why hasn't this been fixed in years?

**Comment by hds** [comment]
@lennartkloock There are a few more details on #2448 about this issue. But it's basically impossible to fix with the current API. The reason is that when a filter is applied, it cannot distinguish the case where enabled has been called by itself from when enabled has been called as part of the normal `if enabled then record-event` flow.

There may be ways to fix this, but it would require a substantial API rework that may require breaking changes.

**Comment by abonander** [comment]
Adding the following comment here for posterity: https://github.com/tokio-rs/tracing/issues/2448#issuecomment-4130431047

> If this can't be fixed without major changes to `tracing`'s internals, then I recommend `enabled!()` be deprecated with a message pointing to this issue or else this is going to keep biting people. I almost introduced the exact same bug as [launchbadge/sqlx#3751](https://github.com/launchbadge/sqlx/issues/3751) in https://github.com/ClickHouse/clickhouse-rs/pull/390 because I was not aware of this until I was reminded by a comment on the former in my notifications.

**Comment by lennartkloock** [comment]
> If this can't be fixed without major changes to `tracing`'s internals, then I recommend `enabled!()` be deprecated with a message pointing to this issue or else this is going to keep biting people. I almost introduced the exact same bug as [launchbadge/sqlx#3751](https://github.com/launchbadge/sqlx/issues/3751) in [ClickHouse/clickhouse-rs#390](https://github.com/ClickHouse/clickhouse-rs/pull/390) because I was not aware of this until I was reminded by a comment on the former in my notifications.

I agree. That would make it clear that the function doesn't work as expected.

**Comment by Aaron1011** [comment]
What about changing the implementation of `enabled!` to give up and produce a false-positive `true` if it would otherwise need to go through the problematic filtering logic?

**Comment by sdigawilkins** [comment]
I threw a workaround up at https://github.com/transact-rs/sqlx/issues/3751#issuecomment-4131222109

I'm sure its a small runtime cost because you'll dispatch the event regardless of whether one of your layers cares about it or not, but at least you get all your events.
